Latest changes in spigot (snakeyaml 1.19) break EssentialsX

Spigot: latest build Spigot-95dd097-c3047a5
EssentialsX: latest jenkins build #541

----------------------------------------------------------------------------------------

Alter last rebuild of spigot EssentialsX produces this exception:

> [11:02:48] [Craft Scheduler Thread - 1/WARN]: [Essentials] Plugin Essentials v2.0.1-bSNAPSHOT generated an exception while executing task 10
> org.yaml.snakeyaml.constructor.ConstructorException: Can't construct a java object for tag:yaml.org,2002:com.earth2me.essentials.settings.Spawns; exception=java.lang.NoSuchFieldException: typeDefinitions
>  in 'reader', line 1, column 1:
>     spawns:
>     ^
> 
>         at org.yaml.snakeyaml.constructor.Constructor$ConstructYamlObject.construct(Constructor.java:313) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at org.yaml.snakeyaml.constructor.BaseConstructor.constructObjectNoCheck(BaseConstructor.java:207) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at org.yaml.snakeyaml.constructor.BaseConstructor.constructObject(BaseConstructor.java:196) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at org.yaml.snakeyaml.constructor.BaseConstructor.constructDocument(BaseConstructor.java:161) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at org.yaml.snakeyaml.constructor.BaseConstructor.getSingleData(BaseConstructor.java:147) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at org.yaml.snakeyaml.Yaml.loadFromReader(Yaml.java:524) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at org.yaml.snakeyaml.Yaml.load(Yaml.java:467) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at com.earth2me.essentials.storage.YamlStorageReader.load(YamlStorageReader.java:41) ~[?:?]
>         at com.earth2me.essentials.storage.AbstractDelayedYamlFileReader.run(AbstractDelayedYamlFileReader.java:33) ~[?:?]
>         at org.bukkit.craftbukkit.v1_12_R1.scheduler.CraftTask.run(CraftTask.java:71) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at org.bukkit.craftbukkit.v1_12_R1.scheduler.CraftAsyncTask.run(CraftAsyncTask.java:52) [spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_152]
>         at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_152]
>         at java.lang.Thread.run(Thread.java:748) [?:1.8.0_152]
> Caused by: org.yaml.snakeyaml.error.YAMLException: java.lang.NoSuchFieldException: typeDefinitions
>         at com.earth2me.essentials.storage.BukkitConstructor$ConstructBukkitMapping.constructJavaBean2ndStep(BukkitConstructor.java:231) ~[?:?]
>         at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.construct(Constructor.java:149) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at com.earth2me.essentials.storage.BukkitConstructor$ConstructBukkitMapping.construct(BukkitConstructor.java:217) ~[?:?]
>         at org.yaml.snakeyaml.constructor.Constructor$ConstructYamlObject.construct(Constructor.java:308) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         ... 13 more
> Caused by: java.lang.NoSuchFieldException: typeDefinitions
>         at java.lang.Class.getDeclaredField(Class.java:2070) ~[?:1.8.0_152]
>         at com.earth2me.essentials.storage.BukkitConstructor$ConstructBukkitMapping.constructJavaBean2ndStep(BukkitConstructor.java:224) ~[?:?]
>         at org.yaml.snakeyaml.constructor.Constructor$ConstructMapping.construct(Constructor.java:149) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         at com.earth2me.essentials.storage.BukkitConstructor$ConstructBukkitMapping.construct(BukkitConstructor.java:217) ~[?:?]
>         at org.yaml.snakeyaml.constructor.Constructor$ConstructYamlObject.construct(Constructor.java:308) ~[spigot-1.12.2.jar:git-Spigot-95dd097-c3047a5]
>         ... 13 more

When i rollback to previous spigot build (Spigot-5695bca-53fccdf) everything works fine.

--- a/Essentials/src/com/earth2me/essentials/storage/BukkitConstructor.java
+++ b/Essentials/src/com/earth2me/essentials/storage/BukkitConstructor.java
@@ -9,17 +9,11 @@ import org.bukkit.enchantments.Enchantment;
 import org.bukkit.inventory.ItemStack;
 import org.bukkit.material.MaterialData;
 import org.bukkit.plugin.Plugin;
-import org.yaml.snakeyaml.TypeDescription;
-import org.yaml.snakeyaml.constructor.Constructor;
 import org.yaml.snakeyaml.constructor.CustomClassLoaderConstructor;
-import org.yaml.snakeyaml.error.YAMLException;
-import org.yaml.snakeyaml.introspector.Property;
+import org.yaml.snakeyaml.introspector.PropertyUtils;
 import org.yaml.snakeyaml.nodes.*;
 
-import java.lang.reflect.Field;
-import java.util.List;
 import java.util.Locale;
-import java.util.Map;
 
 
 public class BukkitConstructor extends CustomClassLoaderConstructor {
@@ -30,6 +24,10 @@ public class BukkitConstructor extends CustomClassLoaderConstructor {
         this.plugin = plugin;
         yamlClassConstructors.put(NodeId.scalar, new ConstructBukkitScalar());
         yamlClassConstructors.put(NodeId.mapping, new ConstructBukkitMapping());
+
+        PropertyUtils propertyUtils = getPropertyUtils();
+        propertyUtils.setSkipMissingProperties(true);
+        setPropertyUtils(propertyUtils);
     }
 
 
@@ -170,7 +168,6 @@ public class BukkitConstructor extends CustomClassLoaderConstructor {
         }
     }
 
-
     private class ConstructBukkitMapping extends ConstructMapping {
         @Override
         public Object construct(final Node node) {
@@ -215,103 +212,6 @@ public class BukkitConstructor extends CustomClassLoaderConstructor {
                 return new Location(world, x, y, z, yaw, pitch);
             }
             return super.construct(node);
-        }
-
-        @Override
-        protected Object constructJavaBean2ndStep(final MappingNode node, final Object object) {
-            Map<Class<? extends Object>, TypeDescription> typeDefinitions;
-            try {
-                final Field typeDefField = Constructor.class.getDeclaredField("typeDefinitions");
-                typeDefField.setAccessible(true);
-                typeDefinitions = (Map<Class<? extends Object>, TypeDescription>) typeDefField.get(BukkitConstructor.this);
-                if (typeDefinitions == null) {
-                    throw new NullPointerException();
-                }
-            } catch (Exception ex) {
-                throw new YAMLException(ex);
-            }
-            flattenMapping(node);
-            final Class<? extends Object> beanType = node.getType();
-            final List<NodeTuple> nodeValue = node.getValue();
-            for (NodeTuple tuple : nodeValue) {
-                ScalarNode keyNode;
-                if (tuple.getKeyNode() instanceof ScalarNode) {
-                    // key must be scalar
-                    keyNode = (ScalarNode) tuple.getKeyNode();
-                } else {
-                    throw new YAMLException("Keys must be scalars but found: " + tuple.getKeyNode());
-                }
-                final Node valueNode = tuple.getValueNode();
-                // keys can only be Strings
-                keyNode.setType(String.class);
-                final String key = (String) constructObject(keyNode);
-                try {
-                    Property property;
-                    try {
-                        property = getProperty(beanType, key);
-                    } catch (YAMLException e) {
-                        continue;
-                    }
-                    valueNode.setType(property.getType());
-                    final TypeDescription memberDescription = typeDefinitions.get(beanType);
-                    boolean typeDetected = false;
-                    if (memberDescription != null) {
-                        switch (valueNode.getNodeId()) {
-                            case sequence:
-                                final SequenceNode snode = (SequenceNode) valueNode;
-                                final Class<? extends Object> memberType = memberDescription.getListPropertyType(key);
-                                if (memberType != null) {
-                                    snode.setListType(memberType);
-                                    typeDetected = true;
-                                } else if (property.getType().isArray()) {
-                                    snode.setListType(property.getType().getComponentType());
-                                    typeDetected = true;
-                                }
-                                break;
-                            case mapping:
-                                final MappingNode mnode = (MappingNode) valueNode;
-                                final Class<? extends Object> keyType = memberDescription.getMapKeyType(key);
-                                if (keyType != null) {
-                                    mnode.setTypes(keyType, memberDescription.getMapValueType(key));
-                                    typeDetected = true;
-                                }
-                                break;
-                        }
-                    }
-                    if (!typeDetected && valueNode.getNodeId() != NodeId.scalar) {
-                        // only if there is no explicit TypeDescription
-                        final Class<?>[] arguments = property.getActualTypeArguments();
-                        if (arguments != null) {
-                            // type safe (generic) collection may contain the
-                            // proper class
-                            if (valueNode.getNodeId() == NodeId.sequence) {
-                                final Class<?> t = arguments[0];
-                                final SequenceNode snode = (SequenceNode) valueNode;
-                                snode.setListType(t);
-                            } else if (valueNode.getTag().equals(Tag.SET)) {
-                                final Class<?> t = arguments[0];
-                                final MappingNode mnode = (MappingNode) valueNode;
-                                mnode.setOnlyKeyType(t);
-                                mnode.setUseClassConstructor(true);
-                            } else if (property.getType().isAssignableFrom(Map.class)) {
-                                final Class<?> ketType = arguments[0];
-                                final Class<?> valueType = arguments[1];
-                                final MappingNode mnode = (MappingNode) valueNode;
-                                mnode.setTypes(ketType, valueType);
-                                mnode.setUseClassConstructor(true);
-                            } else {
-                                // the type for collection entries cannot be
-                                // detected
-                            }
-                        }
-                    }
-                    final Object value = constructObject(valueNode);
-                    property.set(object, value);
-                } catch (Exception e) {
-                    throw new YAMLException("Cannot create property=" + key + " for JavaBean=" + object + "; " + e.getMessage(), e);
-                }
-            }
-            return object;
         }
     }
 }

